sql/core/src/main/scala/org/apache/spark/sql/DataFrameNaFunctions.scala

```diff
@@ -468,37 +481,62 @@ final class DataFrameNaFunctions private[sql](df: DataFrame) {
       s"Unsupported value type ${v.getClass.getName} ($v).")
   }
 
+  private def getTargetType[T](value: T): AbstractDataType = {
+    value match {
+      case _: Double | _: Long => NumericType
+      case _: String => StringType
+      case _: Boolean => BooleanType
+      case _ =>
+        throw new IllegalArgumentException(
+          s"Unsupported value type for fill(): ${value.getClass.getName} ($value).")
+    }
+  }
+
+  private def typeMatches(targetType: AbstractDataType, sourceType: DataType): Boolean = {
+    (targetType, sourceType) match {
+      case (NumericType, dt) => dt.isInstanceOf[NumericType]
+      case (StringType, dt) => dt == StringType
+      case (BooleanType, dt) => dt == BooleanType
+      case _ =>
+        throw new IllegalArgumentException(s"$targetType is not matched for fill().")
+    }
+  }
+
   /**
    * Returns a new `DataFrame` that replaces null or NaN values in specified
    * numeric, string columns. If a specified column is not a numeric, string
    * or boolean column it is ignored.
    */
   private def fillValue[T](value: T, cols: Seq[String]): DataFrame = {
```

Comment:
I think we should have a single entrance of this functionality: `def fillValue[T](value: T, cols: Seq[Attribute])`, and caller side should resolve columns if they only have names.

Reply:
Thanks for the suggestion!